scheduler/plugin/pool.go

```diff
@@ -19,124 +19,131 @@ package wasm
 import (
 	"context"
 	"sync"
-
-	"k8s.io/apimachinery/pkg/types"
 )
 
 // guestPool manages guest to pod assignments in a scheduling or binding cycle.
-type guestPool[guest any] struct {
+//
+// Assumptions made about the lifecycle are taken from the below diagram
+// https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/#extension-points
+type guestPool[guest comparable] struct {
 	// newGuest is a function to create a new guest.
 	newGuest func(context.Context) (guest, error)
 
-	// pool is a pool of unused guest instances.
-	pool sync.Pool
-
-	// schedulingPodUID is the UID of the pod that is being scheduled.
-	// The plugin updates this field at the beginning of each scheduling cycle (PreFilter).
-	//
-	// Note: We cannot unassign the guest instance in PostFilter/Permit
-	// because PostFilter is not enough to notice failures in the scheduling cycle,
-	// it's called only when the Pod is rejected in PreFilter or Filter.
-	// So, we need to trach Pods in the scheduling cycle and the binding cycle separately
-	// so that we can know which guest instance to unassign at PreFilter.
-	schedulingPodUID types.UID
-
-	// assignedToSchedulingPod has the guest instance assignedToSchedulingPod to the pod.
-	// assignedToSchedulingPod instance won't be put back to the pool until the scheduling cycle of this Pod is finished.
-	// The plugin will update this field at the beginning of each scheduling cycle (PreFilter) along with schedulingPodUID.
-	assignedToSchedulingPod guest
-
-	// assignedToSchedulingPodLock is a lock to protect the access to the assigned instance.
-	// The scheduler may call Filter(), AddPod(), RemovePod() concurrently, and we need to take a lock.
-	// But, other methods of the plugin should be invoked for the same Pod serially.
-	assignedToSchedulingPodLock sync.Mutex
-
-	// assignedToBindingPod has the guest instances for Pods in binding cycle.
-	assignedToBindingPod map[types.UID]guest
-}
+	mux sync.RWMutex
```

Comment:
so this iteration uses a simple lock for now. This is under the assumption that access to one plugin isn't concurrent enough to warrant the additional complexity of read and write locks, where often you need a write one. We can later optimize this, but for now, simple IMHO is better.